### PR TITLE
Remove dependency on transforms3d

### PR DESCRIPTION
**Pull request by adeak**
This reimplements rotation around axis and reflection from transforms3d, see https://github.com/pyvista/pyvista/issues/1388.

I've added some additional input checks, even though the two functions are not documented. ~~There's a pair of tests that compare the old and new inputs using `hypothesis`.~~ With float32 inputs there might be numerical differences, so this is not necessarily the final state of the functions, just a baseline to compare against when deciding to remove the dependency on `trasnforms3d`.

<strike>I've kept the original implementations for now so we can test the new implementations against them. To compare the two versions:
```sh
pytest -v -k 'test_tmp_transforms3d_correct'
```
</strike>

Things to fix before finalizing this:

- [x] remove old implementations
- [x] remove sanity check test
- [x] remove transforms3d dependency
- [x] consider edge cases
- [x] explain math in docstrings

### CODE REVIEW

**Comment by akaszynski** [comment]
I'll review this once you're ready.  I use this in a few closed source projects that are transform dependent; they'll most likely identify edge cases.

**Comment by adeak** [comment]
Thanks. I'll only get to it if there's a decision on the issue to use our own implementation that isn't copied over.

**Comment by leycec** [comment]
As the innocent bystander who packaged PyVista for Gentoo, I'm generally all-in for maximally compact dependency trees. Thanks for the great work, @adeak!

**Comment by adeak** [comment]
@akaszynski  in light of https://github.com/pyvista/pyvista/issues/1388#issuecomment-865114619 what do you think we should do? Copy the code from transforms3d, or rewrite from scratch?

**Comment by akaszynski** [comment]
> Copy the code from transforms3d, or rewrite from scratch?

Copy then rewrite.  That way you have something to start from, and we optimize it as needed.

**Comment by adeak** [comment]
OK, but I already have something to start from ;)

**Comment by akaszynski** [comment]
> OK, but I already have something to start from ;)

Up to you then and whatever is the easiest.  Would be great to get faster transforms as I think there's some speed improvements to be made.

**Comment by adeak** [comment]
Well this PR has code that should be fully compatible. I can compare performances. But are you sure the speed issues are with _creating_ the transformations? I'd expect that applying them (which is beyond the scope of transforms3d) will always take a lot more time.

**Comment by akaszynski** [comment]
> Well this PR has code that should be fully compatible. I can compare performances. But are you sure the speed issues are with creating the transformations? I'd expect that applying them (which is beyond the scope of transforms3d) will always take a lot more time.

You're correct, that's out of scope of this PR.  Speed shouldn't be an issue then (sorry, should have looked at the diff).

**Comment by whophil** [comment]
As I was the one who originally brought in `transforms3d` I feel like I should chime in here. I am in support of removing the dependency now that we have an idea for how little of it is actually in use.

The idea was to make it easy to apply other transformations. But that doesn't necessarily need to be a part of PyVista, as the user can build their own transformation matrices as they see fit and apply them using the `DataSetFilters.transform()` filter. (On performance: This filter is the bottleneck in performance, not anything in generating transform matrices. If I recall correctly, we considered applying all the point and field transformations ourselves in NumPy, but ran into issues with edge cases and found it better to stick with VTK's implementation at the time).

This PR prompted me to take a look around `transformations.py` again. I wonder if we might also consider removing or deprecating a few other functions.

- `transformations.apply_transformation_to_points()` is used only by
- `helpers.axis_rotation()` which is only used by
- `DataSet.rotate_x()`, `DataSet.rotate_y()` and `DataSet.rotate_z()`

These are methods for rotating the _points only_ of a dataset around the x/y/z axes. They are also very fast because we leverage NumPy broadcasting to do the rotations. But `rotate_x()` and friends could be rewritten to use `axis_angle_rotation()` and to feed the resulting matrix to `DataSetFilters.transform()`. This would give better continuity of interface and behavior between `rotate_x()` and the other transformation filters, such as `DataSetFilters.reflect()`, likely at the cost of performance.

I believe @akaszynski uses these `rotate_*()` methods in closed source projects so he may have some opinions about how they should function.

**Comment by adeak** [comment]
> As I was the one who originally brought in `transforms3d` I feel like I should chime in here.

Thanks for the context @whophil, it's much appreciated.

>  (On performance: This filter is the bottleneck in performance, not anything in generating transform matrices. If I recall correctly, we considered applying all the point and field transformations ourselves in NumPy, but ran into issues with edge cases and found it better to stick with VTK's implementation at the time)

This sounds interesting to me, I'll try to look into this sometime. I can't speak for the design part of the problem; I'm OK with any outcome.

**Comment by adeak** [comment]
I've finished the refactor. I know the two functions aren't part of the documented API, but I've improved the docstrings anyway. Perhaps they should be properly documented?

The coverage was alright, but I've added dedicated tests for each function to make them complete.

This is ready for review.

File: pyvista/utilities/transformations.py [hunk 1]
```diff
@@ -1,20 +1,214 @@
...
+
+    """
     if deg:
+        # convert to radians
         angle *= np.pi / 180
-    import transforms3d as tf3d
-    return tf3d.axangles.axangle2aff(axis, angle, point=point)
+
+    # return early for no rotation
+    if np.isclose(angle % (2 * np.pi), [0, 2 * np.pi], rtol=0, atol=1e-10).any():
```
Comment (akaszynski) [inline pyvista/utilities/transformations.py:84]: Good check for this.

**Review by akaszynski — APPROVED** [review]
I see no issues with this, and I commend adding full documentation to non-public methods.

Thanks for adding this; it's nice to have one less dependency.

**Comment by akaszynski** [comment]
> the two functions aren't part of the documented API, but I've improved the docstrings anyway

You explain the math and that's sufficient.  It's not going to be rendered in sphinx so I'm happy with it.

**Comment by adeak** [comment]
OK, thanks for the review.

**Comment by akaszynski** [comment]
> The coverage was alright, but I've added dedicated tests for each function to make them complete.

The more I work on projects (big and small), the more I realize how important it is to have dedicated tests for each method.  One of our issues is we don't run tests in order from isolated --> integrated (unit --> functional) and when we break something small, it's harder to pointpoint the test.  We might be able to order tests one day, but for now this works.

File: pyvista/utilities/transformations.py (hunk 1 above)
Comment (whophil) [reply pyvista/utilities/transformations.py:84]: Is this desirable? I spend a lot of time in optimization world where gradients are preferably smooth. This if statement introduces a flat  "notch" in the gradient for `angle` = +/- 1e-10. A quick test shows that there is no actual singularity in the transform at `angle = 0`

Comment (akaszynski) [reply pyvista/utilities/transformations.py:84]: @whophil, I didn't consider this.  @adeak, we may consider dropping this.

Comment (adeak) [reply pyvista/utilities/transformations.py:84]: @whophil what do you mean by "gradients"? This is just an attempt to check that the rotation angle is `n * 2 * pi`, in which case we don't have to do anything. For all intents and purposes a rotation with 1e-10 degrees is not a rotation at all, especially since VTK uses single precision for points if I recall correctly.

This is purely a performance optimization, so we can just remove it or be more strict with `atol`. I just don't (yet) understand the situation where this breaks something.

Comment (whophil) [reply pyvista/utilities/transformations.py:84]: Sorry bad explanation. All I mean is that I would expect all elements of the transform matrix (and point transformations using it) to be smooth w.r.t. angle. With the if statement, I don't think that will be true.

Regarding precision, the output of this function could conceivably be used on non-VTK (and therefore double prec) points, so an assumption about precision seems out of place.

Comment (adeak) [reply pyvista/utilities/transformations.py:84]: I see, thanks for clarifying.

Indeed, the true rotation is smooth, and with the if this is not technically the case. But we're working with technology where
```py
>>> 0.1 + 0.2 == 0.3
False
```

And where 21 turns are no longer a complete turn
```py
>>> (21 * 2*np.pi) % (2 * np.pi)
7.105427357601002e-15
```

Although the function is for internal use only, fair enough about potential more general applicability. Even if it were only to be used for single precision, I haven't done any rigorous testing with precision. For what it's worth I think I honed in one the 1e-10 value when I set up `hypothesis` tests to compare results of the original transformations and the reimplemented ones.

But again, if there's fears of this leading to wrong results (fears that outweight the fears of computing Rodrigues' formula for no reason for identical cases) then we can just remove the fast path.

Comment (whophil) [reply pyvista/utilities/transformations.py:84]: If you remove the fast path, then a rotation angle of 0.0 does not give you an exact identity matrix. Despite this, I prefer this behavior over the built-in "fast path" because the output is smooth, and I can set up the "fast path" outside of the function call if need be. I could certainly be an exception.

For most applications I doubt that the "fast path" saves that much time because time is mostly spent in application of the transformation itself and not in building the transform matrix.

Comment (adeak) [reply pyvista/utilities/transformations.py]: Unlike the above examples, `sin(0) == 0` and `cos(0) == 1` are reliable truths. So no, without the fast path we still get an exact identity matrix for 0 angle. As I said, this is purely a performance optimization.

Partly why I went with this design is because _some people_ raised performance concerns early on :P And I just stuck with it. I think it makes no practical difference. I'll just change it to exact comparison with 0, which will still give us some true positives, without affecting near-zero values. Done in https://github.com/pyvista/pyvista/pull/1392/commits/eae56be3cb0b8bc5458b42c6de62378210bbad23.
